fltk/util/data_sampler_utils.py

```diff
@@ -0,0 +1,232 @@
+                else:
+                    others = [grp for grp in range(n_labels) if grp != group]
+                    if random.choice(others) == group_id:
+                        if group_clients[counter] == client_id:
+                            indices.append(sample_idx)
+                        counter = (counter + 1) % len(group_clients)
+
+        labels = [dataset.targets[i] for i in indices]
+        logging.info("nr of samplers in client with rank {}: {}".format(rank, len(indices)))
+        logging.info("distribution in client with rank {}: {}".format(rank, Counter(labels)))
+
+        random.seed(seed+client_id)  # give each client a unique shuffle
+        random.shuffle(indices) # shuffle indices to spread the labels
+        
+        self.indices = indices
+
+    def __iter__(self) -> Iterator[int]:
+        return iter(self.indices)
+
+    def __len__(self) -> int:
+        logging.info("length of client with rank {} : {}".format(self.rank, len(self.indices)))
+        return len(self.indices)
+
+class DirichletSampler(DistributedSampler):
+    def __init__(self, dataset, rank, world_size, alpha=0.5, seed=42): 
+        self.rank = rank
+        super(DirichletSampler, self).__init__(
+            dataset, world_size, rank, False)
+
+        client_id = rank - 1
+        n_clients = world_size - 1
+
+        # order the indices by label
+        ordered_by_label = [[] for i in range(len(dataset.classes))]
+        for index, target in enumerate(dataset.targets):
+            ordered_by_label[target].append(index)
+
+        n_labels = len(ordered_by_label)
+
+        np.random.seed(seed)
+        indices = []
+        for labels in ordered_by_label:
+            n_samples = len(labels)
+            allocation = np.random.dirichlet([alpha]*n_clients)*n_samples
+            allocation = allocation.astype(int)
+            start_index = allocation[0:client_id].sum()
+            end_index = 0
+            if client_id + 1 == n_clients: # last client
+                end_index = n_samples 
+            else:
+                end_index = start_index + allocation[client_id]
+
+            selection = labels[start_index:end_index]
+            indices.extend(selection)
+
+        labels = [dataset.targets[i] for i in indices]
+        logging.info("nr of samplers in client with rank {}: {}".format(rank, len(indices)))
+        logging.info("distribution in client with rank {}: {}".format(rank, Counter(labels)))
+
+        random.seed(seed+client_id)  # give each client a unique shuffle
+        random.shuffle(indices) # shuffle indices to spread the labels
+        
+        self.indices = indices
+
+    def __iter__(self) -> Iterator[int]:
+        return iter(self.indices)
+
+    def __len__(self) -> int:
+        logging.info("length of client with rank {} : {}".format(self.rank, len(self.indices)))
+        return len(self.indices)
+
+
+if __name__ == "__main__":
```

Comment:
Is this used for testing or something else?

Reply:
Only for testing, will remove this.